### ISSUE TITLE
Generic oAuth2 login provider is unusable

### ISSUE THREAD

**Issue by MrGeneration** [zammad/zammad#2951]
### Infos:

* Used Zammad version: 3.2 and earlier
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any

This issue is a dedicated split from issue https://github.com/zammad/zammad/issues/2944 to remain best possible overview.

### Desired change:

We will be removing the generic oAuth2 authentication entirely from Zammad.
This does not affect other authentication providers.

I'd like to have this change fairly quickly (reasons below) - currently I personally am for 3.5 **if possible**.

Please let me know early enough when we will change this. It requires documentation changes and heads up on our community just in case anybody uses it! We might also have to inform paying customers (internal discussion required).

### Scope:

The may affect a very small number of Zammad users that have fiddled around the existing issues with the generic authentication to make it work.

We believe that removing it is for the better. It will reduce confusion and maintainability for the code as this functionally technically is broken (and never really worked).

This also affects these issues that need re-evaluation if they are still current or would need to be closed because of the removal.

 * https://github.com/zammad/zammad/issues/1431
 * https://github.com/zammad/zammad/issues/868

### Reason for this change:

The generic oAuth2 has been confusing community users regulary.
During that time we also came to the point that this functionality is broken. We also learned that there's no such thing like a generical oAuth integration - it's technically a nightmare.

Most oAuth authentications behave differently from each other, even though they're following the same standards. This thus requires Zammad to have third party specific integrations for the affected oAuth-Providers when ever needed.

We also indirectly came to that point on the community as well (for reference):
 * https://community.zammad.org/t/scope-field-on-oauth2-generic-setup/3288
 * https://community.zammad.org/t/nextcloud-as-oauth2-provider-not-working-minimum-one-identifier-login-firstname-lastname-phone-or-email-for-user-is-required/3737/4?u=mrgeneration

**Comment by mwmahlberg** [comment]
If I might add my 2 cents as a prospective user: While I can understand the change from a developers perspective, I am strongly against it from a user's perspective.

OAUTH(2)/OIDC are ubiquitous and removing support for them would limit the number of potential adopters significantly. That is less than ideal for a "freemium" business model, where you want to have as many adopters as possible, and especially the ones that need single sign on solutions, since those are the ones most likely to become paying customers. Hence removing support for them because it does not work at the moment is the wrong decision, imho.

Obviously, you can judge your _current_ user base much better than I can, but frankly, the lack of OAUTH(2)/OIDC and a less than stable SAML authentication (see [my according post in the forum](https://community.zammad.org/t/saml-oidc-third-party-authentication/2533/39?u=mwmahlberg)) would prevent adoption of the most likely paying customers in the future; it sure does for at least one company I know ;) .
Basically, without proper support for at least one of the most used SSO protocols, I doubt Zammad will make it even into the evaluation phase.

Please do not take this post as criticism or complaint, but as it is meant: 2 cents of a prospective user.

**Comment by MrGeneration** [comment]
Maybe this got a bit off.
This issue is about removing *one* of the many oAuth2 authentication options that are available in Zammad. The generic oAuth2 authentication is *not working* as of now.

We had to learn there is no such thing like "generic oAuth2" which is why this functionality is technically rubbish and more causing confusion than use. 

I mean don't get me wrong, if you have a solution on how to provide generic oAuth authentication that is working without issues and are capable of maintaining it, that would be great. As of now I don't see it as a working functionality which we can fix.

**Comment by mwmahlberg** [comment]
@MrGeneration I dare to object. There are various authentication flows, mappers etc. Most vendors take the approach to give a few examples for the most common auth servers, say Keycloak and Spring Boot, get into great detail on the mappers required and the settings suggested. Since after all, zammad will be only one client in an authentication domain, the client settings _can_ be highly specific to zammad without interfering with other clients.

Sadly, my Ruby is close to non-existent, otherwise I would offer to do a PR. If I can do anything else furthering this matter, I'd be happy to.

**Comment by MrGeneration** [comment]
Thanks for your time and patience.
So what you're basically proposing would be a mapping functionality within Zammad that would allow Zammad to technically convert all incoming keys it every could get in any way from the auth source?

Especially providing several configuration samples is -most of the times- a very time consuming and (sadly) sometimes out of scope. Personally I'm no big fan or growing a documentation too much, because it will loose it's main focus and start to confuse in worst cases.

**Comment by thorsteneckel** [comment]
The three main incompatibilities I've seen and experienced so far that are blocking a fully generic OAuth2 object are:

- Combined Fields  eg. remote `name` to Zammads `firstname`, `lastname`
- Nested/advanced structures: eg. remote `Array(email_address)` or `Hash[info, Hash[email_address, Array(email_address)]]` to Zammads `email_address`
- The need to perform (multiple) follow up requests: e.g. 1st request just authenticates without a response payload, second request fetches actual user data (as for Google)

All of those would require a major overhead on Zammads side to create an abstraction that will never be able to fulfil all cases. On the other hand Omniauth, the gem framework for handling OAuth authentication, in combination with the extremely simple syntax of Rub provides essentially the same. I can only encourage you to read the documentation. Is very simple. I've seen providers with less than 5 lines of code.

However, to get back on track let me summarise the situation: Zammad currently offers the "generic OAuth" authentication provider (next to Gitlab, Office365, Facebook, etc.). This generic provider is broken by design which means that it never worked and will never work. It is based on the generic Omniauth gem which is just the template to inherit from and for building an authentication gem on top of it. That's the point where it's unsuitable for our current implementation: We can't build a gem at runtime. We just can't because it's technically not possible.
To summarise further the "generic OAuth" is and was broken and will never work. Therefore we remove it. This is what issue is for.

I see you point @mwmahlberg and I like your spirit wanting to contribute! I also agree that we should evaluate if Zammad could provide a dedicated, new "generic OAuth" in a way that you described. Just a very simple direct mapping. However, it would not cover the edge cases I described at the beginning of this issue. For this cases programming a Ruby gem/class is much more efficient in terms of development and maintenance costs.
However, we currently can't provide this because the technical fundament is not present and our roadmap is focused on more urgently needed/requested features. Therefore we can/should create a dedicated post over at the [community board](http://community.zammad.org/) where we collect example use cases. Once we got enough examples and requests we can destil/summarise the requirements into an issue. Currently we got rough ideas and single use cases descriptions but that's not enough for an issue.

Thanks for your understanding! Looking forward to subscribe to the/your community tread!
